# E-Commerce Analysis using olist Data

### Data Retrieval

In [1]:
%load_ext sql
import os
import psycopg2
import matplotlib.pyplot as plt
import numpy as np
from sqlalchemy import create_engine

# Connection details
host = "localhost"
database = "olist"
user = "postgres"
password = "test"
db_url = f"postgresql://{user}:{password}@{host}/{database}"

# Connect to database
engine = create_engine(db_url)
%sql engine

In [2]:
%%sql --save table_names --no-execute
SELECT TABLE_NAME
FROM INFORMATION_SCHEMA.TABLES
WHERE TABLE_TYPE = 'BASE TABLE' AND TABLE_CATALOG = 'olist' 
      AND TABLE_SCHEMA = 'public'

Running query in 'postgresql://postgres:***@localhost/olist'

Skipping execution...

In [3]:
table_names = %sql SELECT * FROM table_names
table_name_list = list(table_names.DataFrame()['table_name'])

Generating CTE with stored snippets: 'table_names'

Running query in 'postgresql://postgres:***@localhost/olist'

9 rows affected.

In [4]:
table_name_list

['geo_location',
 'customers',
 'orders',
 'order_payments',
 'order_reviews',
 'order_items',
 'products',
 'sellers',
 'product_translation']

In [5]:
%sqlcmd explore --table order_items

In [6]:
%sqlcmd columns --table sellers

+----------------+-------------+----------+---------+---------------+---------+
|      name      |     type    | nullable | default | autoincrement | comment |
+----------------+-------------+----------+---------+---------------+---------+
|   seller_id    | VARCHAR(50) |  False   |   None  |     False     |   None  |
| seller_zipcode |   CHAR(5)   |   True   |   None  |     False     |   None  |
|  seller_city   | VARCHAR(50) |   True   |   None  |     False     |   None  |
|  seller_state  |  VARCHAR(5) |   True   |   None  |     False     |   None  |
+----------------+-------------+----------+---------+---------------+---------+

In [7]:
%sqlcmd columns --table geo_location

+-----------+------------------+----------+---------+---------------+---------+
|    name   |       type       | nullable | default | autoincrement | comment |
+-----------+------------------+----------+---------+---------------+---------+
|  zipcode  |     CHAR(5)      |   True   |   None  |     False     |   None  |
|  latitude | DOUBLE_PRECISION |   True   |   None  |     False     |   None  |
| longitude | DOUBLE_PRECISION |   True   |   None  |     False     |   None  |
|    city   |   VARCHAR(50)    |   True   |   None  |     False     |   None  |
|  geostate |     CHAR(2)      |   True   |   None  |     False     |   None  |
+-----------+------------------+----------+---------+---------------+---------+

In [8]:
%%sql --save orders_raw --no-execute
WITH avg_seller_coors
       AS (SELECT AVG(latitude) AS seller_latitude, AVG(longitude) AS seller_longitude, seller_zipcode
             FROM sellers AS s
             LEFT JOIN geo_location AS gl ON s.seller_zipcode = gl.zipcode
            GROUP BY seller_zipcode), 
    seller_locations
       AS (SELECT *
             FROM sellers AS s
             LEFT JOIN avg_seller_coors AS gl ON s.seller_zipcode = gl.seller_zipcode),
    product_items_sellers
       AS (SELECT *
             FROM (SELECT *
                     FROM order_items AS oi
                     LEFT JOIN seller_locations AS s ON oi.seller_id = s.seller_id) AS ois
             LEFT JOIN products AS p ON ois.product_id = p.product_id),
    product_items_sellers_translated
       AS (SELECT * 
             FROM product_items_sellers AS p
             LEFT JOIN product_translation AS pt ON p.product_category = pt.category),
    avg_customer_coors
       AS (SELECT AVG(latitude) AS customer_latitude, AVG(longitude) AS customer_longitude, customer_zipcode
             FROM customers AS c
             LEFT JOIN geo_location AS gl ON c.customer_zipcode = gl.zipcode
            GROUP BY customer_zipcode),
    customer_locations
       AS (SELECT *
             FROM customers AS c
             LEFT JOIN avg_customer_coors AS gl ON c.customer_zipcode = gl.customer_zipcode)
SELECT *
  FROM orders AS o
  LEFT JOIN customer_locations AS cloc ON o.customer_id = cloc.customer_id
  LEFT JOIN product_items_sellers_translated AS pi ON o.order_id = pi.order_id
  LEFT JOIN order_payments AS op ON o.order_id = op.order_id
  LEFT JOIN order_reviews AS oe ON o.order_id = oe.order_id

Running query in 'postgresql://postgres:***@localhost/olist'

Skipping execution...

In [9]:
orders_raw_sql = %sql SELECT * FROM orders_raw
orders = orders_raw_sql.DataFrame()

Generating CTE with stored snippets: 'orders_raw'

Running query in 'postgresql://postgres:***@localhost/olist'

119143 rows affected.

In [10]:
orders = orders.loc[:,~orders.columns.duplicated()].copy()

### Data Preprocessing

In [11]:
# Check missing values
orders.isna().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase                   0
order_approved                 177
order_delivered_carrier       2086
order_delivered_customer      3421
order_estimated_delivery         0
customer_unique_id               0
customer_zipcode                 0
customer_city                    0
customer_state                   0
customer_latitude              322
customer_longitude             322
order_item_id                  833
product_id                     833
seller_id                      833
shipping_limit_date            833
price                          833
freight_value                  833
seller_zipcode                 833
seller_city                    833
seller_state                   833
seller_latitude               1098
seller_longitude              1098
product_category              2542
product_name_length           2542
product_desc_length           2542
product_photos_qty  

#### Missing Order Delivery Dates

In [12]:
orders[orders["order_delivered_carrier"].isnull()]["order_status"].value_counts()

order_status
canceled       670
unavailable    652
invoiced       378
processing     376
created          5
approved         3
delivered        2
Name: count, dtype: int64

"order_delivered_carrier" shows the order posting timestamp, i.e. when it was handed to the logistic partner. As seen, the missing values come from mostly canceled and unavailable orders, as they should be. The other portion is mostly invoiced and processing orders, as the orders are still in progress, so they are also to be expected. The 2 orders, that is recorded to be delivered, need to be investigated further:

In [13]:
orders[(orders["order_delivered_carrier"].isnull()) & (orders["order_status"] == "delivered")]

order_id                       customer_id  \
9098    2aa91108853cecb43c84a5dc5b277475  afeb16c7f46396c0ed54acb45ccaaa40   
100172  2d858f451373b04fb5c984a1cc2defaf  e08caf668d499a6d643dafd7c5cc498a   

       order_status      order_purchase      order_approved  \
9098      delivered 2017-09-29 08:52:58 2017-09-29 09:07:16   
100172    delivered 2017-05-25 23:22:43 2017-05-25 23:30:16   

       order_delivered_carrier order_delivered_customer  \
9098                       NaT      2017-11-20 19:44:47   
100172                     NaT                      NaT   

       order_estimated_delivery                customer_unique_id  \
9098                 2017-11-14  a2ac81ecc3704410ae240e74d4f0af40   
100172               2017-06-23  d77cf4be2654aa70ef150f8bfec076a6   

       customer_zipcode  ... payment_sequential payment_type  \
9098              13334  ...                1.0  credit_card   
100172            91330  ...                1.0  credit_card   

        payment_installments  payment_value                         review_id  \
9098                     1.0         193.98  e945d1831a3d98008913fc31dcbb804d   
100172                   4.0         194.00  4e755f114e50d33b9ac6a56e0d7d3ea9   

       review_score review_title review_comment  review_create  \
9098            5.0         None           None     2017-10-17   
100172          5.0         None           None     2017-06-25   

             review_answer  
9098   2017-10-17 10:56:02  
100172 2017-06-27 01:49:04  

[2 rows x 45 columns]

These 2 orders can be removed from the dataset due to unreliable / missing delivery dates.

In [14]:
orders = orders[~orders["order_id"].isin(["2d858f451373b04fb5c984a1cc2defaf", "2aa91108853cecb43c84a5dc5b277475"])].copy()

Only 7 delivered orders with missing customer delivery dates can be imputed with estimated delivery for consistency:

In [15]:
orders["order_delivered_customer"] = orders["order_delivered_customer"].fillna(orders["order_estimated_delivery"])

In [16]:
orders[orders["order_approved"].isnull()]["order_status"].value_counts()

order_status
canceled     157
delivered     15
created        5
Name: count, dtype: int64

For calculation of order processing time, the missing order_approved times can be imputed with order_purchase_timestamp:

In [17]:
orders['order_approved'] = orders['order_approved'].fillna(orders['order_purchase'])

In [18]:
orders.isna().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase                   0
order_approved                   0
order_delivered_carrier       2084
order_delivered_customer         0
order_estimated_delivery         0
customer_unique_id               0
customer_zipcode                 0
customer_city                    0
customer_state                   0
customer_latitude              322
customer_longitude             322
order_item_id                  833
product_id                     833
seller_id                      833
shipping_limit_date            833
price                          833
freight_value                  833
seller_zipcode                 833
seller_city                    833
seller_state                   833
seller_latitude               1098
seller_longitude              1098
product_category              2542
product_name_length           2542
product_desc_length           2542
product_photos_qty  

#### Missing Customer Locations

Some customer zip codes are not present in the geo_location table, which results in 322 orders with missing customer locations. These locations can be imputed with the median coordinates for the city that the customer is in:

In [19]:
median_coords = orders.groupby('customer_city')[['customer_latitude', 'customer_longitude']].transform('median')

orders[['customer_latitude', 'customer_longitude']] = orders[['customer_latitude', 'customer_longitude']].fillna(median_coords)

In [20]:
len(orders[orders['customer_latitude'].isnull()])

53

In [21]:
len(orders[orders['customer_latitude'].isnull()]['customer_city'].unique())

45

As seen, there are still 53 orders with 45 different customer cities for which there are no entries in the provided geolocation table. Since there are many different cities and only 53 missing orders, it is not worth adding the coordinates for each city. These 53 orders can be removed from the dataframe:

In [22]:
orders = orders[~orders['customer_latitude'].isnull()].copy()

#### Missing Seller Locations

In [23]:
orders[(~orders['seller_city'].isnull()) & (orders['seller_longitude'].isnull())]['seller_city'].value_counts()

seller_city
sao paulo          140
pocos de caldas     69
brasilia            42
curitiba             9
aruja                3
porto alegre         2
Name: count, dtype: int64

In [24]:
median_seller_coords = orders.groupby('seller_city')[['seller_latitude', 'seller_longitude']].transform('median')

orders[['seller_latitude', 'seller_longitude']] = orders[['seller_latitude', 'seller_longitude']].fillna(median_seller_coords)

#### Missing Order Items

In [25]:
orders.isna().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase                   0
order_approved                   0
order_delivered_carrier       2082
order_delivered_customer         0
order_estimated_delivery         0
customer_unique_id               0
customer_zipcode                 0
customer_city                    0
customer_state                   0
customer_latitude                0
customer_longitude               0
order_item_id                  832
product_id                     832
seller_id                      832
shipping_limit_date            832
price                          832
freight_value                  832
seller_zipcode                 832
seller_city                    832
seller_state                   832
seller_latitude                832
seller_longitude               832
product_category              2541
product_name_length           2541
product_desc_length           2541
product_photos_qty  

In [26]:
orders[orders['order_item_id'].isnull()]['order_status'].value_counts()

order_status
unavailable    645
canceled       179
created          5
invoiced         2
shipped          1
Name: count, dtype: int64

In [27]:
print(f"Missing order items percentage: {np.round(100*len(orders[orders['order_item_id'].isnull()]) / len(orders), 2)}%")

Missing order items percentage: 0.7%


As seen, there are 832 orders with missing information on order items, most of which seem to be due to the order being unavailable or canceled. Since these orders count for only 0.7% of all orders, they can be removed from the dataset:

In [28]:
orders = orders[~orders['order_item_id'].isnull()].copy()

In [29]:
orders.isna().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase                   0
order_approved                   0
order_delivered_carrier       1251
order_delivered_customer         0
order_estimated_delivery         0
customer_unique_id               0
customer_zipcode                 0
customer_city                    0
customer_state                   0
customer_latitude                0
customer_longitude               0
order_item_id                    0
product_id                       0
seller_id                        0
shipping_limit_date              0
price                            0
freight_value                    0
seller_zipcode                   0
seller_city                      0
seller_state                     0
seller_latitude                  0
seller_longitude                 0
product_category              1709
product_name_length           1709
product_desc_length           1709
product_photos_qty  

The 20 orders with missing product size information can be removed as well:

In [30]:
orders = orders[~orders['product_length_cm'].isnull()].copy()

#### Missing Category Names

In [31]:
orders.isna().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase                   0
order_approved                   0
order_delivered_carrier       1251
order_delivered_customer         0
order_estimated_delivery         0
customer_unique_id               0
customer_zipcode                 0
customer_city                    0
customer_state                   0
customer_latitude                0
customer_longitude               0
order_item_id                    0
product_id                       0
seller_id                        0
shipping_limit_date              0
price                            0
freight_value                    0
seller_zipcode                   0
seller_city                      0
seller_state                     0
seller_latitude                  0
seller_longitude                 0
product_category              1690
product_name_length           1690
product_desc_length           1690
product_photos_qty  

As seen, the number of missing product categories is not the same as the missing categories coming from the translation table. That is due to the fact that there are missing translations for the following categories:

In [32]:
orders[orders['product_category'] != orders['category']]['product_category'].value_counts()

product_category
portateis_cozinha_e_preparadores_de_alimentos    15
pc_gamer                                         10
Name: count, dtype: int64

In [33]:
cat_translate = {'portateis_cozinha_e_preparadores_de_alimentos': 'portable_kitchen_food_preparers', 'pc_gamer': 'pc_gamer'}
orders.loc[orders['product_category'] != orders['category'], 'category_translation'] = orders.loc[orders['product_category'] != orders['category'], 'product_category'].map(cat_translate)

In [34]:
# Delete the additional category column
orders = orders.drop('category', axis=1).reset_index(drop=True)

The remaining 1690 orders with unknown product categories can be imputed with "unknown" (or 0 for numerical values) to not lose information:

In [35]:
orders['product_category'] = orders['product_category'].fillna('unknown')
orders['category_translation'] = orders['category_translation'].fillna('unknown')
orders['product_name_length'] = orders['product_name_length'].fillna(0)
orders['product_desc_length'] = orders['product_desc_length'].fillna(0)
orders['product_photos_qty'] = orders['product_photos_qty'].fillna(0)

#### Missing payment information

In [36]:
orders[orders['payment_type'].isna()]

order_id                       customer_id  \
29537  bfbd0f9bdef84302105ad712db648a6c  86dc2ffce2dfff336de2f386a786e574   
29538  bfbd0f9bdef84302105ad712db648a6c  86dc2ffce2dfff336de2f386a786e574   
29539  bfbd0f9bdef84302105ad712db648a6c  86dc2ffce2dfff336de2f386a786e574   

      order_status      order_purchase      order_approved  \
29537    delivered 2016-09-15 12:16:38 2016-09-15 12:16:38   
29538    delivered 2016-09-15 12:16:38 2016-09-15 12:16:38   
29539    delivered 2016-09-15 12:16:38 2016-09-15 12:16:38   

      order_delivered_carrier order_delivered_customer  \
29537     2016-11-07 17:11:53      2016-11-09 07:47:38   
29538     2016-11-07 17:11:53      2016-11-09 07:47:38   
29539     2016-11-07 17:11:53      2016-11-09 07:47:38   

      order_estimated_delivery                customer_unique_id  \
29537               2016-10-04  830d5b7aaa3b6f1e9ad63703bec97d23   
29538               2016-10-04  830d5b7aaa3b6f1e9ad63703bec97d23   
29539               2016-10-04  830d5b7aaa3b6f1e9ad63703bec97d23   

      customer_zipcode  ... payment_sequential payment_type  \
29537            14600  ...                NaN         None   
29538            14600  ...                NaN         None   
29539            14600  ...                NaN         None   

       payment_installments  payment_value                         review_id  \
29537                   NaN            NaN  6916ca4502d6d3bfd39818759d55d536   
29538                   NaN            NaN  6916ca4502d6d3bfd39818759d55d536   
29539                   NaN            NaN  6916ca4502d6d3bfd39818759d55d536   

      review_score review_title  \
29537          1.0         None   
29538          1.0         None   
29539          1.0         None   

                                       review_comment  review_create  \
29537  nao recebi o produto e nem resposta da empresa     2016-10-06   
29538  nao recebi o produto e nem resposta da empresa     2016-10-06   
29539  nao recebi o produto e nem resposta da empresa     2016-10-06   

            review_answer  
29537 2016-10-07 18:32:28  
29538 2016-10-07 18:32:28  
29539 2016-10-07 18:32:28  

[3 rows x 44 columns]

As seen, there are only 3 orders with missing payment information that is likely caused by entry errors. These orders can also be removed from the dataset for completeness:

In [37]:
orders = orders[~orders['payment_type'].isnull()].copy().reset_index(drop=True)

In [38]:
orders.isna().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase                   0
order_approved                   0
order_delivered_carrier       1251
order_delivered_customer         0
order_estimated_delivery         0
customer_unique_id               0
customer_zipcode                 0
customer_city                    0
customer_state                   0
customer_latitude                0
customer_longitude               0
order_item_id                    0
product_id                       0
seller_id                        0
shipping_limit_date              0
price                            0
freight_value                    0
seller_zipcode                   0
seller_city                      0
seller_state                     0
seller_latitude                  0
seller_longitude                 0
product_category                 0
product_name_length              0
product_desc_length              0
product_photos_qty  

The only missing values remaining are:

- order_delivered_carrier = mostly due to canceled or ongoing orders, will be kept in the dataset as is.
- Reviews = There are 978 orders with no reviews, meaning the customers did not provide any rating, which is to be expected in real-world data.
- For the reviews that do exist, there are titles or comments missing, that is also acceptable and realistic.

In [39]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 118233 entries, 0 to 118232
Data columns (total 44 columns):
 #   Column                    Non-Null Count   Dtype         
---  ------                    --------------   -----         
 0   order_id                  118233 non-null  object        
 1   customer_id               118233 non-null  object        
 2   order_status              118233 non-null  object        
 3   order_purchase            118233 non-null  datetime64[ns]
 4   order_approved            118233 non-null  datetime64[ns]
 5   order_delivered_carrier   116982 non-null  datetime64[ns]
 6   order_delivered_customer  118233 non-null  datetime64[ns]
 7   order_estimated_delivery  118233 non-null  datetime64[ns]
 8   customer_unique_id        118233 non-null  object        
 9   customer_zipcode          118233 non-null  object        
 10  customer_city             118233 non-null  object        
 11  customer_state            118233 non-null  object        
 12  cu

So the preprocessed dataset has 118233 orders with information across 44 columns, ready for an exploratory data analysis.

In [42]:
orders.to_csv(os.path.join('../data/processed/', 'processed_orders.csv'), index=False, date_format='%Y-%m-%d %H:%M:%S')